### ISSUE TITLE
Fix `rule-selector-property-disallowed-list` false positives for nesting selectors

### ISSUE THREAD

**Issue by emmacharp** [stylelint/stylelint#7488]
### What minimal example or steps are needed to reproduce the bug?

```css
.search {
   & input {
      width: 100%;
   }
}
```

### What minimal configuration is needed to reproduce the bug?

```json 
{ 
  "rules": { 
    "rule-selector-property-disallowed-list": [{ "/^((?!input).)*$/": "width" }]
  }
}
```

### How did you run Stylelint?

```shell
stylelint *.css
```

[Online demo](https://stylelint.io/demo/#N4Igxg9gJgpiBcIB0BnGBDATmAFgAmAB0A7PPAMjwEtiAHAVwBcCSyyB3KqRnePARgAMggKQBuPAHoAVHhQ4I9ADZQ8AIxh4IAawA06pnnRhG9dEqUBPPJhi0ImRjFXTJrPAF8SX4iVQZsfBoGZiJSPE5uXgFhcSlZYghmW3tHZ30dPFdvEj80LFw8ADMHAFsWcMiePiFRCRkbOwcnKAztLLdiDxBdECKqJRgAOXRSuEQYAA9R2kGkMBQUHvAIYn6AcwQQMLJCEExlGBQ9vh22PYPBgFo0QZMHK9pMCFoYR0srqCoUcyUIdmcVyU30YJzwAG1gHg9pIAHoACnhAH4AITBJgASiQGOkABI3AhoSAqjg9p4ALruHzdXqQNZUdYAMTK6EYWwAVihVstYLQlogznsUIxLINgcRQYS9kpWUdJd4QB4gA)

### Which Stylelint-related dependencies are you using?

```json
{
   "stylelint": "16.2.0"
}
```

### What did you expect to happen?

The nested block should not generate an error.

### What actually happened?

`.search` is evaluated and an error is thrown.

### Do you have a proposal to fix the bug?

Don't evaluate nested selectors... though I'm not sure about collateral damage.

**Comment by ybiquitous** [comment]
@emmacharp Thanks for the report with a reproduction. To clarify, do you want to disallow a `width` prop within all `input` selectors?

I created the [online demo](https://stylelint.io/demo/#N4Igxg9gJgpiBcIB0BnGBDATmAFgAmAB0A7PPAMjwEtiAHAVwBcCSyyB3KqRnePARgAMggKQBuVngC+JGcRKoM2fDQbMipPJ268Bw8bJIgANCABmVADYwAcugC2cRDAAeD2taRgUKE+AjEFgDmCCAaZIQgmPTWKJF84WyR0dYAtGjWYIwQmKm0mBC0MJiMAJ6pUFQo6JaWEOwwUKmWVYzxeADawHiRqkztkdo8kdIAupJyUn6QgVRBAGI59uiMoQBWKAF+sLS+iImRKGXWLcRtCD0gliswR5GyIFJAA), which doesn't seem to be a problem. See the config:

```json
{
  "rules": {
    "rule-selector-property-disallowed-list": [{ "input": "width" }]
  }
}
```

**Comment by emmacharp** [comment]
The contrary, actually. The example is about disallowing the `width` property within any selector *not* containing `input`. But since the `input` selector is nested inside a `.search` selector, an error results.

My regex in the example was bunk, sorry. If you try your demo with this one instead, you should see what I mean:
`/^((?!input).)*$/`

As always, thanks for the help!

**Comment by ybiquitous** [comment]
Ah, I understand it now. See the new [demo](https://stylelint.io/demo/#N4Igxg9gJgpiBcIB0BnGBDATmAFgAmAB0A7PPAMjwEtiAHAVwBcCSyyB3KqRnePARgAMggKQBuPAHoAVHhQ4I9ADZQ8AIxh4IAawA06pnnRhG9dEqUBPPJhi0ImRjFXTJrPAF8SX4iVQZsfBoGZiJSPE5uXgFhcSlZYghmW3tHZ30dPFdvEj80LFw8ADMHAFsWcMiePiFRCRkbOwcnKAztLLdiDxBdECKqJRgAOXRSuEQYAA9R2kGkMBQUHvAIYn6AcwQQMLJCEExlGBQ9vh22PYPBgFo0QZMHK9pMCFoYR0srqCoUcyUIdmcVyU30YJzwAG1gHg9pIAHoACnhAH4AITBJgASiQGOkABI3AhoSAqjg9p4ALruHzdXqQNZUdYAMTK6EYWwAVihVstYLQlogznsUIxLINgcRQYS9kpWUdJd4QB4gA).

```css
.search {
  & input {
    width: 100%; /* should be ok, but actually reported */
 /* ^^^^^ Unexpected property "width" for selector ".search" */
  }
}

.search input {
  width: 100%; /* not reported, ok */
}

.search form {
  width: 100%; /* reported, ok */
/*^^^^^ Unexpected property "width" for selector ".search form" */
}
```
I've labeled the issue as **ready to implement**. Please consider [contributing](https://stylelint.io/contributing) if you have time.

There are [steps on how to fix a bug in a rule](https://stylelint.io/developer-guide/rules#fix-a-bug-in-a-rule) in the Developer guide.
```json
{
  "rules": {
    "rule-selector-property-disallowed-list": [{ "/^((?!input).)*$/": "width" }]
  }
}
```

This should be a bug of `rule-selector-property-disallowed-list` for nested rules.

I've labeled the issue as **ready to implement**. Please consider [contributing](https://stylelint.io/contributing) if you have time.

There are [steps on how to fix a bug in a rule](https://stylelint.io/developer-guide/rules#fix-a-bug-in-a-rule) in the Developer guide.

**Comment by ybiquitous** [comment]
I've updated this issue body to include the correct reproduction and the online demo.

**Comment by emmacharp** [comment]
Thanks for rephrasing in a standard way. I'll see what I can do today & tomorrow.

**Comment by ybiquitous** [comment]
We can use [`postcss-resolve-nested-selector`](https://www.npmjs.com/package/postcss-resolve-nested-selector) to resolve a nested selector, which is installed already.

Otherwise, we may be able to install and use [`@csstools/selector-resolve-nested`](https://www.npmjs.com/package/@csstools/selector-resolve-nested) instead. The package follows the CSS Nesting spec, while `postcss-resolve-nested-selector` doesn't.

For details, see https://github.com/stylelint/stylelint/issues/6234#issuecomment-1902790274

**Comment by emmacharp** [comment]
Well...

I have never worked with Postcss before and the truth is I'm more a designer than a programmer. Heh. So figuring out what's going on is proving a bit challenging. 

The way I understand it, we need to prevent walking through child nodes' rules. I'm not sure how the plugins mentioned above can help doing that. Am i missing something? 

Not quite sure about collateral damage either. Do you all think about something that could be broken by preventing the rule to walk through child nodes in the AST?

**Comment by ybiquitous** [comment]
Ah, I see. Don't mind. Adding `@csstools/selector-resolve-nested` was started with PR #7496, so let's wait for the job. This issue should be addressed more easily if the job is done.

**Comment by emmacharp** [comment]
@ybiquitous, do we still wait PR #7496 to be closed to progress here? If not, I'd try to see what I can do for it to be included in 16.3.0

Thanks!

**Comment by ybiquitous** [comment]
In this issue, I favor installing `@csstools/selector-resolve-nested` because #7496 tries replacing `postcss-resolve-nested-selector` with `@csstools/selector-resolve-nested` but `rule-selector-property-disallowed-list` doesn't rely on `postcss-resolve-nested-selector`. There is no breaking change when using `@csstools/selector-resolve-nested`.

CC: @romainmenke if you notice something, feel free to comment.

> If not, I'd try to see what I can do for it to be included in 16.3.0

Honestly, I'd like to release 16.3.0 as soon as possible because the version will include some bug fixes.

So, you don't need to rush, we can release 16.3.1 soon when resolving this issue.

**Comment by romainmenke** [comment]
I think that actual issue might simpler to resolve :)
But maybe not.

I'll open a PR with a proposed fix.
